### PR TITLE
entropy stats

### PR DESCRIPTION
**Pull request by williamberman**
This adds three main sets of statistics for the entropy of predicted pixel token distributions.

The stats all are broken categorically into buckets of the range of percent masked pixels. As we don't have formal timesteps determining the amount of noise this seemed to be the cleanest dividing line for collections with similar entropy. All stats are hardcoded to use 10 buckets -- bucket 0: 0-0.1 bucket 1: 0.1-0.2, etc...

1) We break the cross entropy w.r.t truth out by buckets in addition to the standard loss which is averaged over all images
2) We plot the average entropy per bucket -- i.e. the entropy for the distribution of each pixel is calculated and then all pixel entropies w/in a bucket are averaged together
3) We plot the full distribution of one pixel per bucket via a histogram and rug chart

### CODE REVIEW

File: muse/training_utils.py [hunk 1]
```diff
@@ -19,7 +19,9 @@
 from typing import Any, Dict, Iterable, Optional, Union
 
 import numpy as np
+import pandas as pd
```
Comment (patil-suraj) [inline muse/training_utils.py:22]: let's add this in  `setup.py` as requirement.

File: muse/training_utils.py [hunk 2]
```diff
@@ -291,3 +293,163 @@ def load_state_dict(self, state_dict: dict) -> None:
...
+    pixel_probs = F.softmax(logits, dim=-1)
+    pixel_probs[~masked_tokens] = 0
+    image_probs_numerator = pixel_probs.sum(-2)
+    image_probs = image_probs_numerator / num_masked_pixels
+
+    image_log_probs = image_probs.log()
+
+    entropy_per_image = -((image_probs * image_log_probs).sum(-1))
+
+    total_buckets = 10
```
Comment (patil-suraj) [inline muse/training_utils.py:338]: really cool!

File: training/train_muse.py [hunk 3]
```diff
@@ -23,9 +23,9 @@
 from typing import Any, List, Tuple, Union
 
 import numpy as np
+import plotly.express as px
```
Comment (patil-suraj) [inline training/train_muse.py:26]: let's also add this as dep as we are adding this  to training script.

File: training/train_muse.py [hunk 4]
```diff
@@ -642,6 +645,43 @@ def prepare_inputs_and_labels(
...
+                        global_step + 1,
+                    )
+
+                if (
+                    ("log_token_probability_distributions_every" in config.experiment)
+                    and ((global_step + 1) % config.experiment.log_token_probability_distributions_every == 0)
+                    and accelerator.is_main_process
+                ):
+                    log_token_probability_distributions(logits, input_ids, mask_id, accelerator, global_step + 1)
+
```
Comment (patil-suraj) [inline training/train_muse.py:684]: Looks good !

**Review by patil-suraj — APPROVED** [review]
Looks great! Just left couple of comments, feel free to merge after that :)

File: training/train_muse.py (hunk 3 above)
Comment (williamberman) [reply training/train_muse.py:26]: done!

File: muse/training_utils.py (hunk 1 above)
Comment (williamberman) [reply muse/training_utils.py:22]: done!
